executor/insert_common.go

```diff
@@ -470,6 +471,45 @@ func insertRowsFromSelect(ctx context.Context, base insertCommon) error {
 	memUsageOfRows := int64(0)
 	memUsageOfExtraCols := int64(0)
 	memTracker := e.memTracker
+
+	var (
+		dataBatch [][]types.Datum
+		dataCh    chan [][]types.Datum
+	)
+	etlConcurrency, etlBatchSize := sessVars.ETLConcurrency, sessVars.ETLBatchSize
+	if etlMode && etlConcurrency > 0 {
+		workers := createFromSelectWorkers(ctx, base, etlConcurrency)
+		if len(workers) == 0 {
+			// fallback to single thread.
+			etlMode = false
+		} else {
+			affectedRow := uint64(0)
+			dataBatch = make([][]types.Datum, 0, etlBatchSize)
+			dataCh = make(chan [][]types.Datum, 2*etlConcurrency)
```

Comment:
`2*etlConcurrency` may make the reader wait, how about a larger value like 1024 if the memory usage is acceptable?

Reply:
The reader should be fast enough, `2*etlConcurrency` controls the memory usage.